diff --git a/build-llvm.sh b/build-llvm.sh
index 4d76f08..b301760 100755
--- a/build-llvm.sh
+++ b/build-llvm.sh
@@ -16,7 +16,7 @@
 
 set -e
 
-: ${LLVM_VERSION:=llvmorg-10.0.0}
+: ${LLVM_VERSION:=release/11.x}
 ASSERTS=OFF
 BUILDDIR=build
 unset HOST
@@ -182,6 +182,6 @@ cmake \
     $CMAKEFLAGS \
     ..
 
-$BUILDCMD ${CORES+-j$CORES} install/strip
+$BUILDCMD -j7 install/strip
 
 cp ../LICENSE.TXT $PREFIX
