diff --git a/configure b/configure
index 767fb7b..ce4d069 100755
--- a/configure
+++ b/configure
@@ -80,7 +80,7 @@ prefix=${prefix-/usr/local}
 exec_prefix=${exec_prefix-'${prefix}'}
 bindir=${bindir-'${exec_prefix}/bin'}
 libdir=${libdir-'${exec_prefix}/lib'}
-sharedlibdir=${sharedlibdir-'${libdir}'}
+sharedlibdir=${sharedlibdir-'${exec_prefix}/bin'}
 includedir=${includedir-'${prefix}/include'}
 mandir=${mandir-'${prefix}/share/man'}
 shared_ext='.so'
@@ -111,7 +111,7 @@ gcc=0
 warn=0
 debug=0
 old_cc="$CC"
-old_cflags="$CFLAGS"
+old_ldflags="$LDFLAGS"
 OBJC='$(OBJZ)'
 PIC_OBJC='$(PIC_OBJZ)'
 INSTALLTARGETS="install-shared install-static"
@@ -253,6 +253,7 @@ fi
 
 STATICLIB=${LIBNAME}.a
 MAPNAME=${LIBNAME2}.map
+IMPORTLIB=
 
 # extract zlib version numbers from zlib.h
 if test $compat -eq 0; then
@@ -359,25 +360,24 @@ if test "$gcc" -eq 1 && ($cc $CFLAGS -c $test.c) >> configure.log 2>&1; then
   CYGWIN* | Cygwin* | cygwin*)
         ARFLAGS="rcs"
         SFLAGS="${CFLAGS}"
+        LDFLAGS="-L. ${old_ldflags}"
         shared_ext='.dll'
         sharedlibdir='${bindir}'
-        if test $compat -eq 0; then
-          SHAREDLIB=cygz-ng$shared_ext
-        else
-          SHAREDLIB=cygz$shared_ext
-        fi
+        SHAREDLIB='cygz.dll'
         SHAREDLIBM=''
         SHAREDLIBV=''
         SHAREDTARGET=$SHAREDLIB
-        IMPORTLIB="${LIBNAME}.dll.a"
+        IMPORTLIB='libz.dll.a'
         LDSHARED=${LDSHARED-"$cc"}
-        LDSHAREDFLAGS="-shared -Wl,--out-implib,${IMPORTLIB},--version-script,${SRCDIR}/${MAPNAME}"
-        LDSHAREDLIBC=""
+        LDSHARED=${LDSHARED-"$cc -shared -Wl,-export-all -Wl,--enable-auto-image-base -Wl,--out-implib=${IMPORTLIB}"}
+        LDSHAREDLIBC=''
         DEFFILE='win32/${LIBNAME2}.def'
         RC="${CROSS_PREFIX}windres"
         RCFLAGS='--define GCC_WINDRES'
         RCOBJS='zlibrc.o'
-        STRIP="${CROSS_PREFIX}strip"
+        STRIP="${CROSS_PREFIX}strip
+        LDSHARED=${LDSHARED-"$cc -shared -Wl,-export-all -Wl,--enable-auto-image-base -Wl,--out-implib=${IMPORTLIB}"}
+        LDSHAREDLIBC=''
         EXE='.exe' ;;
   MSYS* | msys*)
         ARFLAGS="rcs"
