diff --color -Naur bison-3.7.1-orig/m4/dup2.m4 bison-3.7.1/m4/dup2.m4
--- bison-3.7.1-orig/m4/dup2.m4	2020-08-05 19:01:35.882536100 +0300
+++ bison-3.7.1/m4/dup2.m4	2020-08-05 19:02:08.470325200 +0300
@@ -68,7 +68,7 @@
       [case "$host_os" in
          mingw*) # on this platform, dup2 always returns 0 for success
            gl_cv_func_dup2_works="guessing no" ;;
-         cygwin*) # on cygwin 1.5.x, dup2(1,1) returns 0
+         cygwin* | msys*) # on cygwin 1.5.x, dup2(1,1) returns 0
            gl_cv_func_dup2_works="guessing no" ;;
          aix* | freebsd*)
                  # on AIX 7.1 and FreeBSD 6.1, dup2 (1,toobig) gives EMFILE,
