diff --git a/include/libssh2.h b/include/libssh2.h
index 386bcc892..110e32ae2 100644
--- a/include/libssh2.h
+++ b/include/libssh2.h
@@ -100,7 +100,7 @@ extern "C" {
 /* Allow alternate API prefix from CFLAGS or calling app */
 #ifndef LIBSSH2_API
 # ifdef LIBSSH2_WIN32
-#  ifdef _WINDLL
+#  if defined(_WINDLL) || defined(libssh2_EXPORTS)
 #   ifdef LIBSSH2_LIBRARY
 #    define LIBSSH2_API __declspec(dllexport)
 #   else
